src/Workspaces/Core/Portable/Classification/ClassificationServiceWithProviders.cs

```diff
@@ -0,0 +1,188 @@
+                {
+                    extensionManager.HandleException(provider, e);
+                }
+            }
+
+            return classifiedSpan;
+        }
+
+        public override ImmutableArray<ClassifiedSpan> GetLexicalClassifications(SourceText text, TextSpan span, CancellationToken cancellationToken)
+        {
+            var extensionManager = _workspace.Services.GetService<IExtensionManager>();
+            var spans = SharedPools.Default<List<ClassifiedSpan>>().Allocate();
+            var spanSet = SharedPools.Default<HashSet<ClassifiedSpan>>().Allocate();
+            try
+            {
+                var context = new ClassificationContext(spans, spanSet);
+                foreach (var provider in this.GetProviders())
+                {
+                    try
+                    {
+                        if (!extensionManager.IsDisabled(provider))
+                        {
+                            provider.AddLexicalClassifications(text, span, context, cancellationToken);
+                        }
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        throw;
+                    }
+                    catch (Exception e) when (extensionManager.CanHandleException(provider, e))
+                    {
+                        extensionManager.HandleException(provider, e);
+                    }
+                }
+
+                return spans.ToImmutableArray();
+            }
+            finally
+            {
+                SharedPools.Default<List<ClassifiedSpan>>().ClearAndFree(spans);
+                SharedPools.Default<HashSet<ClassifiedSpan>>().ClearAndFree(spanSet);
+            }
+        }
+
+        public override async Task<ImmutableArray<ClassifiedSpan>> GetSyntacticClassificationsAsync(Document document, TextSpan span, CancellationToken cancellationToken)
+        {
+            var extensionManager = _workspace.Services.GetService<IExtensionManager>();
+            var spans = SharedPools.Default<List<ClassifiedSpan>>().Allocate();
+            var spanSet = SharedPools.Default<HashSet<ClassifiedSpan>>().Allocate();
+            try
+            {
+                var context = new ClassificationContext(spans, spanSet);
+                foreach (var provider in this.GetProviders())
+                {
+                    try
+                    {
+                        if (!extensionManager.IsDisabled(provider))
+                        {
+                            await provider.AddSyntacticClassificationsAsync(document, span, context, cancellationToken).ConfigureAwait(false);
+                        }
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        throw;
+                    }
+                    catch (Exception e) when (extensionManager.CanHandleException(provider, e))
```

Comment:
Yes, there is a helper method already..  But I didn't want to have it create a lambda capture every invocation so I copied the logic directly out of the helper method.

---
In reply to: [78663816](https://github.com/dotnet/roslyn/pull/13741#discussion_r78663816) [](ancestors = 78663816)